addons/portal/mail_message.py

```diff
@@ -19,45 +19,17 @@
-from openerp.tools.translate import _
+from openerp import models
+from openerp.osv import expression
 
+class IrRule(models.Model):
+    _inherit = 'ir.rule'
 
-class mail_message(osv.Model):
-    """ Update of mail_message class, to restrict mail access. """
-    _inherit = 'mail.message'
-
-    def _search(self, cr, uid, args, offset=0, limit=None, order=None,
-        context=None, count=False, access_rights_uid=None):
-        """ Override that adds specific access rights of mail.message, to remove
-            all internal notes if uid is a non-employee
-        """
-        if uid == SUPERUSER_ID:
-            return super(mail_message, self)._search(cr, uid, args, offset=offset, limit=limit, order=order,
-                context=context, count=count, access_rights_uid=access_rights_uid)
-        group_ids = self.pool.get('res.users').browse(cr, uid, uid, context=context).groups_id
-        group_user_id = self.pool.get("ir.model.data").get_object_reference(cr, uid, 'base', 'group_user')[1]
-        if group_user_id not in [group.id for group in group_ids]:
-            args = [('subtype_id', '!=', False)] + list(args)
-
-        return super(mail_message, self)._search(cr, uid, args, offset=offset, limit=limit, order=order,
-            context=context, count=count, access_rights_uid=access_rights_uid)
-
-    def check_access_rule(self, cr, uid, ids, operation, context=None):
-        """ Add Access rules of mail.message for non-employee user:
-            - read:
-                - raise if the type is comment and subtype NULL (internal note)
-        """
-        if uid == SUPERUSER_ID:
-            return super(mail_message, self).check_access_rule(cr, uid, ids=ids, operation=operation, context=context)
-        group_ids = self.pool.get('res.users').browse(cr, uid, uid, context=context).groups_id
-        group_user_id = self.pool.get("ir.model.data").get_object_reference(cr, uid, 'base', 'group_user')[1]
-        if group_user_id not in [group.id for group in group_ids]:
-            cr.execute('SELECT DISTINCT id FROM "%s" WHERE type = %%s AND subtype_id IS NULL AND id = ANY (%%s)' % (self._table), ('comment', ids,))
-            if cr.fetchall():
-                raise AccessError(_('The requested operation cannot be completed due to security restrictions. Please contact your system administrator.\n\n(Document type: %s, Operation: %s)') % \
-                        (self._description, operation))
-
-        return super(mail_message, self).check_access_rule(cr, uid, ids=ids, operation=operation, context=context)
+    def _compute_domain(self, cr, uid, model_name, mode="read"):
+        domain = super(IrRule, self)._compute_domain(cr, uid, model_name, mode=mode)
+        if model_name == 'mail.message' and self.pool['res.users'].has_group(cr, uid, 'base.group_user'):
+            domain = expression.AND([
+                domain,
+                ['&', ('type', '=', 'comment'), ('subtype_id', '=', False)]
+            ])
```

Comment:
@xmo-odoo I think the condition is incorrect. The domain returned by `_compute_domain()` defines which records you have access to, and you cannot access records that satisfy `type='comment' AND subtype_id IS NULL`. Therefore you have to negate this condition in `_compute_domain()`:

    domain = expression.AND([domain, [`|`, ('type', '!=', 'comment'), ('subtype_id', '!=', False)]])

or maybe something simpler, like what `_search()` does:

    domain = expression.AND([domain, [('subtype_id', '!=', False)]])